Preflight Checks: Add Docker build step

vaurdan opened the pull request:
## Description

In order to have high fidelity preflight tests, this PR implements a new preflight check that tries to build a Docker image with a similar process that is used for the production image.

When Docker isn't available on the machine, it will warn the user, recommending to run the preflight tests on a machine with Docker available.

This is intended as a first iteration for a Docker check step, to pave the way for more complex and accurate Docker builds, that could, for instance, take in account the site environment variables, and Node version.

<strike>Currently this check tries to get the node version from a `.nvmrc` file. If this file doesn't exist, or doesn't contain a valid node version, it will fallback to the current version being used to run the preflight checks.</strike>

The Node.JS version is picked with a simple selector, using the [inquirer package](https://www.npmjs.com/package/inquirer). Additionally, to provide a non-interactive alternative, one can use the `--node-version <VERSION>` (or `-n <VERSION`) argument, where `<VERSION>` should be a major Node.JS version - 12, 14 or 16.

![image](https://user-images.githubusercontent.com/5804923/146575945-93d853b7-60eb-4637-bd63-d6fd69e95f92.png)

## Steps to Test

1. Check out PR.
1. Run `npm run build`.
1. `cd` to a VIP Go node site repository
2. Run `node [path-to-vip-go-node-repo]/preflight-checks/dist/index.js`
3. If `docker` is available (and the application builds), you should see the build step succeeding:
```
$ node ../vip-go-node/preflight-checks/dist/index.js
(...)
  [ Step 3/3 ] Building Docker image...
  [ Step 3/3 ] Step details: Trying to build a Docker image for the project, using a production-like build script
  Info: Found Docker Engine 20.10.12
  Info: Using Node.js 16.3.0 to build the image
  Info: Using GitHub repository organization/your-repository
  Info: Building Docker image...
  [ Step 3/3 ] Step successful 👍
```

chriszarate commented on preflight-checks/package.json:
```diff
@@ -8,7 +8,7 @@
   "scripts": {
     "test": "jest",
-    "build": "babel src --out-dir dist",
+    "build": "babel src --out-dir dist --copy-files --no-copy-ignored",
```
What do these flags do?

vaurdan replied:
These flags make sure that the Dockerfile and the build.sh file (in the `src/docker`) directory are copied to the built application in `dist/`. 

https://babeljs.io/docs/en/babel-cli#copy-files

chriszarate commented on preflight-checks/src/checks/03-docker-build.js:
```diff
@@ -0,0 +1,102 @@
+const chalkPackageJson = chalk.yellow( 'package.json' );
+
+const envVariables = {
+	'VIP_GO_APP_ID': '123',
```
Let's also add a few more env vars that we supply and use a clearer "fake" site ID:

```suggestion
	'VIP_GO_APP_ID': 'unknown',
	'VIP_GO_APP_NAME': 'unknown',
	'VIP_GO_APP_ENVIRONMENT': 'local',
```

chriszarate commented on preflight-checks/src/checks/03-docker-build.js:
```diff
@@ -0,0 +1,102 @@
+
+async function findNodeVersion() {
+	// Try to fetch a .nvmrc file
+	const nvmrcPath = await findUp('.nvmrc');
```
This is pretty cool! But since VIP doesn't respect `.nvmrc`, we wouldn't want to push customers in that direction. Maybe a better first iteration would be a simple picker with a hard-coded list of the current LTS major versions:

```
> Please select a Node.js version:
- 16
- 14
- 12
```

chriszarate commented on preflight-checks/src/checks/03-docker-build.js:
```diff
@@ -0,0 +1,102 @@
+		// Check for Node version
+		let nodeVersion = await findNodeVersion();
+		if ( ! nodeVersion ) {
+			console.log( chalk.yellow( '  Warning:' ), `Couldn't infer a valid Node.JS from .nvmrc file. Falling back to local node version.` );
```
Not needed if we implement a picker

chriszarate commented on preflight-checks/src/checks/03-docker-build.js:
```diff
@@ -0,0 +1,102 @@
+		console.log( chalk.blue( '  Info:' ), `Using Node.js ${ chalk.yellow(nodeVersion) } to build the image` );
+
+		// Get REPO_NAME and REPO_ORG
+		const gitConfigPath = await findUp( ".git/config" );
```
The only thing this is used for is to generate the Docker image tag name. I think we can skip all this and alter the script to use a tag that we pass in, maybe something like `vip-preflight-checks:SHA`

chriszarate commented on preflight-checks/src/checks/03-docker-build.js:
```diff
@@ -0,0 +1,102 @@
+		try {
+			console.log( chalk.blue( '  Info:' ), 'Building Docker image...' );
+			await executeShell( `bash ${ __dirname }/../docker/build.sh`, {
+				'REPO_NAME': gitRepo,
+				'REPO_ORG': gitOrg,
```
```suggestion
```

chriszarate commented on preflight-checks/src/docker/build.sh:
```diff
@@ -0,0 +1,74 @@
+
+echo $(pwd);
+
+if [ -z "$REPO_NAME" ] || [ -z "$REPO_ORG" ] || [ -z "$NODE_VERSION" ]; then
+  exit 42
+fi
+
+# Lowercase REPO references. Requires Bash 4 (macOS ships with 3.2).
+IMAGE="${REPO_ORG,,}/${REPO_NAME,,}"
+
+# Tag images with the current commit SHA.
+COMMIT="$(git rev-parse HEAD)"
+IMAGE_TAG="$IMAGE:$COMMIT"
```
```suggestion
# Tag images with the current commit SHA.
COMMIT="$(git rev-parse HEAD)"
IMAGE_TAG="vip-preflight-checks:$COMMIT"
```

chriszarate replied:
Also gets rid of the macOS incompatibility here :)

chriszarate commented:
Left some feedback, this is great so far!

vaurdan commented:
Thank you for the feedback @chriszarate ! I just implemented your suggestions, feel free to take a look.

I also updated the PR description to include the new `--node-version` CLI argument.

elazzabi commented on preflight-checks/package.json:
```diff
@@ -30,6 +30,7 @@
     "command-line-args": "^5.1.1",
     "command-line-usage": "^6.1.1",
     "execa": "^4.0.0",
+    "inquirer": "^6.2.0",
```
We've been using `enquirer` in the [vip-cli](https://github.com/Automattic/vip/blob/develop/package.json#L115) project. I wonder if we should push for using the same dependencies in all our packages? Also, it seems like enquirer uses fewer dependencies 🤔 

What do you think @vaurdan @chriszarate ?

NPM links for reference: [inquirer](https://www.npmjs.com/package/inquirer) | [enquirer](https://www.npmjs.com/package/enquirer)

chriszarate replied:
Standardizing sounds good, as does fewer dependencies :+1:

vaurdan replied:
Should be straight forward to use enquirer instead. I will do the change, thanks for the heads-up! :)

vaurdan commented:
I just pushed a new changeset with the following changes:

 * Using `inquirer` instead of `enquirer`;
 * Run either the `docker-build` or `local-build` check, depending if Docker is available on the local machine.
 * If Docker is available, start the image and run the health check tests against the running container. Otherwise it will run `npm start` locally and run the check using that server.
 * The port can be changed with `--port [PORT]`. By default, it will pick a random port between 3001 and 3999.

@elazzabi @chriszarate feel free to take a look at the new changes. Looking forward for your reviews :) 

( please ignore the duplicated commits in the history. Apparently Github is not hiding amended commits  ¯\\_(ツ)_/¯ )

elazzabi reviewed (approved):
Looks good! Thank you for the changes 🙌
